### PR TITLE
Additional subscription scheme settings

### PR DESCRIPTION
**Pull request by seb86**
This push is in reference to issue #1.

Changes made are simply adding filters to allow the full use of Subscriptions within this extension and allow other extensions to connect with such as [this one](https://github.com/seb86/woocommerce-subscribe-to-all-the-things-signup-trial-add-on).

Also did a version number change.

### CODE REVIEW

**Comment by franticpsyx** [comment]
@seb86 Green light from me - I'll give you some more time as we think about #42 just in case you need to add anything else in v1.1.

**Comment by seb86** [comment]
@franticpsyx Thank you.

File: includes/class-wcsatt-display.php [hunk 1]
```diff
@@ -526,6 +536,43 @@ public static function filter_price_html( $price, $product ) {
...
+
+		$product_type = $product->product_type;
+
+		$product_schemes = get_post_meta( $product->id, '_wcsatt_schemes', true );
+
+		$force_subscription = get_post_meta( $product->id, '_wcsatt_force_subscription', true );
+
+		if ( in_array( $product_type, WCS_ATT()->get_supported_product_types() ) && $product_schemes ) {
+
+			if ( $force_subscription == 'yes' ) {
```
Comment (thenbrent) [inline includes/class-wcsatt-display.php]: This should be a yoda condition, i.e. `'yes' == $force_subscription`, to conform to the [WP coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions).

**Comment by thenbrent** [comment]
@seb86 LGTM. One issue with use of yoda condition but otherwise 👍 

I'd also suggest not updating version numbers in PRs. That can be done at deployment time rather than during development.

File: includes/class-wcsatt-display.php (hunk 1 above)
Comment (seb86) [inline includes/class-wcsatt-display.php]: @thenbrent I obviously still have much to learn. I did not know of this Yoda condition. I will push a quick fix.

**Comment by seb86** [comment]
@thenbrent I changed the version during development of the extension I'm building. I will remember next time to put it back to the original version before committing.

**Comment by seb86** [comment]
@franticpsyx @thenbrent Any chance this will be merged soon?

File: includes/class-wcsatt-display.php [hunk 2]
```diff
@@ -526,6 +536,43 @@ public static function filter_price_html( $price, $product ) {
...
+			if ( 'yes' == $force_subscription ) {
+
+				if ( $product->variation_id > 0 ) {
+
+					$button_text = __( 'Sign Up this Variation', WCS_ATT::TEXT_DOMAIN );
+
+				} else {
+
+					$button_text = get_option( WC_Subscriptions_Admin::$option_prefix . '_add_to_cart_button_text', __( 'Sign Up Now', WCS_ATT
+::TEXT_DOMAIN ) );
```
Comment (thenbrent) [inline includes/class-wcsatt-display.php]: This doesn't need to be on a new line. 😄

File: includes/class-wcsatt-display.php [hunk 3]
```diff
@@ -526,6 +536,43 @@ public static function filter_price_html( $price, $product ) {
...
+
+		$force_subscription = get_post_meta( $product->id, '_wcsatt_force_subscription', true );
+
+		if ( in_array( $product_type, WCS_ATT()->get_supported_product_types() ) && $product_schemes ) {
+
+			if ( 'yes' == $force_subscription ) {
+
+				if ( $product->variation_id > 0 ) {
+
+					$button_text = __( 'Sign Up this Variation', WCS_ATT::TEXT_DOMAIN );
```
Comment (thenbrent) [inline includes/class-wcsatt-display.php]: This button text feels a little too custom to me to include in core. I'd probably instead always use the value of `WC_Subscriptions_Admin::$option_prefix . '_add_to_cart_button_text'` and if the store owner wants custom text for variations, they can use the `'wcsatt_add_to_cart_text'` filter to set that.

**Comment by thenbrent** [comment]
@seb86 sorry for the delay. I've left two small comments, otherwise this looks good to me and can be merged once those are fixed up and @franticpsyx gives it the 👍

**Comment by franticpsyx** [comment]
@seb86 Took a look and basically agree with @thenbrent 's comments. I'll merge and release v1.1 when the points Brent raised are fixed.

I hoped to have some issues reported for v1.1 RC1 before calling v1.1 final - also waited for a last word from you to merge but caught up by work in the meanwhile.

**Comment by seb86** [comment]
OK @thenbrent and @franticpsyx. I will correct this today and get back to you both.

**Comment by seb86** [comment]
Done. Any chance issue #49 could be addressed quickly @franticpsyx before the release?

File: includes/admin/class-wcsatt-admin.php [hunk 4]
```diff
@@ -36,7 +36,7 @@ public static function init() {
 		add_action( 'woocommerce_product_write_panels', __CLASS__ . '::product_write_panel' );
 
 		// Processes and saves the necessary post meta.
-		add_action( 'woocommerce_process_product_meta', __CLASS__ . '::process_product_meta' );
+		add_action( 'woocommerce_process_product_meta', __CLASS__ . '::process_product_meta', 15, 1 );
```
Comment (seb86) [inline includes/admin/class-wcsatt-admin.php]: I also made sure it saves the meta data after WooCommerce has done it's first, not before.

**Comment by franticpsyx** [comment]
@seb86 

I have a feeling that 6bf86da shouldn't be in SATT, but I might be wrong. I'm not familiar with the Sub Shortcodes extension but this piece of code doesn't look like it's making SATT do something for its self that it wouldn't do otherwise. If it's adding SATT support into another plugin, then it should go in there (conditionally on the presence of SATT).

>Any chance issue #49 could be addressed quickly

Yep, will have a look once we decide what to do with 6bf86da and finally merge this.

**Comment by franticpsyx** [comment]
@seb86 please avoid adding irrelevant commits to this PR to keep things focused - 342cf14 is not relevant with "Additional subscription scheme settings" or adding hooks, which was the intent of this PR :)

**Comment by seb86** [comment]
@franticpsyx Should I then remove commit 6bf86da so you can merge. You may be right about adding it to the subscription shortcodes extension instead.

**Comment by franticpsyx** [comment]
Thanks @seb86 - yep, if you can remove 6bf86da I'll have a final look and then merge 👍

**Comment by seb86** [comment]
@franticpsyx Great. :+1:

**Comment by seb86** [comment]
Ready, Set, Merge, Release  :package:

**Comment by franticpsyx** [comment]
@seb86 Tweaked a few things in https://github.com/Prospress/woocommerce-subscribe-all-the-things/commit/07765c19fa5f23346947ed0c3db52370f1f05a21 

To be honest, I don't think this filter was needed - https://github.com/Prospress/woocommerce-subscribe-all-the-things/commit/07765c19fa5f23346947ed0c3db52370f1f05a21#diff-af820c6be360bb1a8e112cfab4d897ddR167

There's already a filter that allows you to change the option description a few lines below. I'd be happy to remove the `wcsatt_single_product_subscription_scheme_price_html` filter if you can use the one that follows.

**Comment by seb86** [comment]
@franticpsyx I will be able to go over it tomorrow.

**Comment by seb86** [comment]
> To be honest, I don't think this filter was needed - 07765c1#diff-af820c6be360bb1a8e112cfab4d897ddR167

@franticpsyx I'm pretty sure I put that filter there so that I could return the sign up fee and trial. So now I need to change the filter in my extension which I will do tomorrow.

**Comment by franticpsyx** [comment]
>I put that filter there so that I could return the sign up fee and trial. So now I need to change the filter in my extension which I will do tomorrow.

@seb86 Can you use this one instead? - https://github.com/Prospress/woocommerce-subscribe-all-the-things/blob/master/includes/class-wcsatt-display.php#L181

I totally missed there was a filter when looking at your diff - perhaps you missed that as well?

**Comment by seb86** [comment]
> @seb86 Can you use this one instead?

@franticpsyx No, because it needs to filter the function `WC_Subscriptions_Product::get_price_string` on [line 167](https://github.com/Prospress/woocommerce-subscribe-all-the-things/blob/master/includes/class-wcsatt-display.php#L167).

**Comment by seb86** [comment]
@franticpsyx I have [changed the filters](https://github.com/seb86/woocommerce-subscribe-to-all-the-things-signup-trial-add-on/commit/402a3db5531672c3bc7f312a0ff38e2af027c3f3) used in my extension to match your tweak.

**Pull request merged**